File: ts/lengthOfLongestSubstring.ts
function lengthOfLongestSubstring(s: string): number {
  let i = 0;
  let j = 1;
  let maxLength = 0;
  const charSet = new Set();
  const str = s.replace(/ /g, '뷁');

  while (i < str.length) {
    if (i >= j) {
      j += 1;
      continue;
    }

    const chars = str.slice(i, j);
    console.log({ i, j, chars });

    if (j > str.length) {
      i += 1;
      j = i + 1;
      continue;
    }

    if (!charSet.has(chars)) {
      maxLength = Math.max(maxLength, chars.length);
      charSet.add(chars);
    }

    if (chars.includes(str[j])) {
      i += 1;
    } else {
      j += 1;
    }
  }

  console.log(charSet);
  return maxLength;
}

export default () => {
  const s = '   ';
  // const s = 'pwwkew';
  // const s = 'bbbbb';
  // const s = 'abcabcbb';
  const result = lengthOfLongestSubstring(s);
  console.log(result);
};

File: rs/src/length_of_longest_substring.rs
use std::collections::HashMap;

fn length_of_longest_substring(s: String) -> i32 {
    let mut max_length = 0;
    let mut start = 0;
    let mut char_indices = HashMap::new();

    for (end, char) in s.chars().enumerate() {
        if let Some(&index) = char_indices.get(&char) {
            start = start.max(index + 1);
        }
        max_length = max_length.max(end - start + 1);
        char_indices.insert(char, end);
    }

    return max_length as i32;
}

pub fn main() {
    let s: String = "abcabcbb".to_string();
    let result = length_of_longest_substring(s);
    println!("{result}");
}
